In [14]:
# Askhsh 19

import random
random.seed(2019)  # for reproducibility
from simple_markov_chain_lib import markov_chain
import statistics as stat
import numpy as np

M = 50 # number of monte-carlo estimations.
pArray = [1/16, 1/6, 1/12]

## Experiment parameters
NArray = [200, 20000]   # number of samples
steps = 20   # the target time

def calcTheoreticalValue(p, initial, final):
    markov_table = np.array([
        [0, 1, 0],
        [0, 2/3, 1/3],
        [p, 1-p, 0]
    ])
    
    P20 = np.linalg.matrix_power(markov_table,20)
    return P20[initial-1][final-1]
    
def simulation(init_probs, initial): #starting for state = initial
    for p in pArray:
        #dictionary for transition matrix.
        markov_table = {
            1: {2: 1.},
            2: {2: 2/3, 3: 1/3},
            3: {1: p, 2: 1-p}
        }
        estimates = {
            200: [],   #N = 200
            20000: []  #N = 20000
        }

        for _ in range(M):
            #construct a markov chain.
            mc = markov_chain(markov_table, init_probs)
            for N in NArray:
                ## Simulation
                counter = 0
                for _ in range(N):
                    mc.start()  # new experiment
                    for _ in range(steps):  mc.move()
                    if mc.running_state == 1:  counter += 1
                estimates[N].append(counter / N)
        print("====================")
        print("p =",p,":")
        print("Mean for N = 200:       ", stat.mean(estimates[200]))
        print("Variance for N = 200:   ", stat.variance(estimates[200]))
        print("Mean for N = 20000:     ", stat.mean(estimates[20000]))
        print("Variance for N = 20000: ", stat.variance(estimates[20000]))
        if (1 in init_probs):
            print("Theoretical mean value: ", calcTheoreticalValue(p , initial, 1))
        else:
            print("Theoretical mean value: ", calcTheoreticalValue(p , initial, 1))

In [15]:
print("---- initial state = 1 ----")
simulation({1: 1.0}, 1)
print("\n")


print("---- initial state = 3 ----")
simulation({3: 1.0}, 3)

---- initial state = 1 ----
p = 0.0625 :
Mean for N = 200:        0.0167
Variance for N = 200:    9.552040816326532e-05
Mean for N = 20000:      0.015533
Variance for N = 20000:  5.220010204081633e-07
Theoretical mean value:  0.01538461538425158
p = 0.16666666666666666 :
Mean for N = 200:        0.0388
Variance for N = 200:    0.00014444897959183675
Mean for N = 20000:      0.040065
Variance for N = 20000:  2.192678571428571e-06
Theoretical mean value:  0.03999999999973111
p = 0.08333333333333333 :
Mean for N = 200:        0.0196
Variance for N = 200:    6.616326530612245e-05
Mean for N = 20000:      0.020286
Variance for N = 20000:  1.0473510204081637e-06
Theoretical mean value:  0.020408163265301697


---- initial state = 3 ----
p = 0.0625 :
Mean for N = 200:        0.016
Variance for N = 200:    6.122448979591836e-05
Mean for N = 20000:      0.015215
Variance for N = 20000:  5.268622448979592e-07
Theoretical mean value:  0.01538461538436527
p = 0.16666666666666666 :
Mean for N = 200

## Άσκηση 19

1) O μέσος όρος των Μ εκτιμήσεων είναι:    
* Για p = 1/16:  
    * Για $Ν = 200$: $\hat{p} = 0.0167$    
    * Για $Ν = 20000$: $\hat{p} = 0.015533$  
* Για p = 1/6:  
    * Για $Ν = 200$: $\hat{p} = 0.0388$    
    * Για $Ν = 20000$: $\hat{p} = 0.040065$  
* Για p = 1/12:  
    * Για $Ν = 200$: $\hat{p} = 0.0196$    
    * Για $Ν = 20000$: $\hat{p} = 0.020286$  
   

2)   
*  για $p=\frac{1}{16}$:  
$\mathbb{P}[X_{20}=1 | X_0=1] = 0.01538461538425158$  
*  για $p=\frac{1}{6}$:   
$\mathbb{P}[X_{20}=1 | X_0=1] = 0.03999999999973111$  
*  για $p=\frac{1}{12}$:   
$\mathbb{P}[X_{20}=1 | X_0=1] = 0.020408163265301697$  

Παρατηρούμε ότι οι θεωρητικά υπολογισμένες τιμές είναι πολύ κοντά  
σε αυτές που υπολογίσαμε με την μέθοδο Monte Carlo.  

3) Η δειγματική διασπορά των Μ εκτιμήσεων είναι:  
* Για p = 1/16:  
    * Για $Ν = 200$: $\hat{p} = 9.552040816326532e-05$    
    * Για $Ν = 20000$: $\hat{p} =  5.220010204081633e-07$  
* Για p = 1/6:  
    * Για $Ν = 200$: $\hat{p} = 0.00014444897959183675$    
    * Για $Ν = 20000$: $\hat{p} = 2.192678571428571e-06$  
* Για p = 1/12:  
    * Για $Ν = 200$: $\hat{p} = 6.616326530612245e-05$    
    * Για $Ν = 20000$: $\hat{p} = 1.0473510204081637e-06$ 

4) Εάν η Μ.Α. ξεκινάει από την κατάσταση 3, οι νέες εκτιμήσεις είναι:  
* Για p = 1/16:  
    * Για $Ν = 200$: $\hat{p} = 0.016$    
    * Για $Ν = 20000$: $\hat{p} = 0.015215$  
* Για p = 1/6:  
    * Για $Ν = 200$: $\hat{p} = 0.0377$    
    * Για $Ν = 20000$: $\hat{p} = 0.040124$  
* Για p = 1/12:  
    * Για $Ν = 200$: $\hat{p} = 0.0188$    
    * Για $Ν = 20000$: $\hat{p} = 0.020349$
    
Παρατηρούμε ότι δεν αλλάζουν ουσιαστικά οι εκτιμήσεις.

In [42]:
# Askhsh 40

p = 0.6
q = 1-p
init_probs = {"0-0": 1}
N = 100000 #number of experiments

markov_table = {
    "0-0": {"15-0": p, "0-15": q},
    "15-0": {"30-0": p, "15-15": q},
    "0-15": {"15-15": p, "0-30": q},
    "30-0": {"40-0": p, "30-15": q},
    "15-15": {"30-15": p, "15-30": q},
    "0-30": {"15-30": p, "0-40": q},
    "40-0": {"GameA": p, "40-15": q},
    "30-15": {"40-15": p, "Deuce": q},
    "15-30": {"Deuce": p, "15-40": q},
    "0-40": {"15-40": p, "GameB": q},
    "40-15": {"GameA": p, "AdvA": q},
    "15-40": {"AdvB": p, "GameB": q},
    "AdvA": {"GameA": p, "Deuce": q},
    "AdvB": {"Deuce": p, "GameB": q},
    "Deuce": {"AdvA": p, "AdvB": q},
    "GameA": {"GameA": 1},
    "GameB": {"GameB": 1}
}

def winner(state):
    return (state == "GameA" or state == "GameB")

counter = 0
mc = markov_chain(markov_table, init_probs)
for _ in range(N):
    mc.start()
    state = ""
    while not(winner(state)):
        mc.move()
        state = mc.running_state
    if (state == "GameA"): #player A won
        counter += 1
       
phat = counter/N
print("Estimation of probability that PlayerA wins a game:", phat)

Estimation of probability that PlayerA wins a game: 0.7357
